Potential rustc bug related to deriving ByteEq

Right off the bat, here's the minimum reproducible example:
Cargo.toml:
```toml
[package]
name = "bug_rep"
version = "0.1.0"
edition = "2021"

[dependencies]
bytemuck = { version = "1.14.0", features = ["derive"] }
```

```rust
#[derive(bytemuck::ByteEq)]
pub struct Magnitude<const BYTES: usize>([u8; 0]);
```
The bug still occurs identically if swapping out `bytemuck::ByteEq` for `bytemuck::ByteHash`.

The error from `cargo check` is:
```
error[E0107]: missing generics for struct `Magnitude`
 --> src/lib.rs:2:12
  |
2 | pub struct Magnitude<const BYTES: usize>([u8; 0]);
  |            ^^^^^^^^^ expected 1 generic argument
  |
note: struct defined here, with 1 generic parameter: `BYTES`
 --> src/lib.rs:2:12
  |
2 | pub struct Magnitude<const BYTES: usize>([u8; 0]);
  |            ^^^^^^^^^ ------------------
help: add missing generic argument
  |
2 | pub struct Magnitude<BYTES><const BYTES: usize>([u8; 0]);
  |                     +++++++

For more information about this error, try `rustc --explain E0107`.
error: could not compile `bug_rep` (lib) due to 2 previous errors
```
Following it's inane suggestion, unsurprisingly, does not fix the issue.

I got this from the rather more reasonable code:
```toml
[package]
name = "magnitude"
version = "0.1.0"
edition = "2021"

[dependencies]
bytemuck = { version = "1.14.0", features = ["derive", "min_const_generics"] }
```

```rust
#[derive(Clone, Copy, bytemuck::Zeroable, bytemuck::Pod, bytemuck::ByteEq, bytemuck::ByteHash)]
#[repr(transparent)]
pub struct Magnitude<const BYTES: usize>([u8; BYTES]);
```


Which gives an identical error message other than that it shows the longer source code.

```
    Checking magnitude v0.1.0 (/home/lily/dev/magnitude)
error[E0107]: missing generics for struct `Magnitude`
 --> src/static_precision/mod.rs:3:12
  |
3 | pub struct Magnitude<const BYTES: usize>([u8; BYTES]);
  |            ^^^^^^^^^ expected 1 generic argument
  |
note: struct defined here, with 1 generic parameter: `BYTES`
 --> src/static_precision/mod.rs:3:12
  |
3 | pub struct Magnitude<const BYTES: usize>([u8; BYTES]);
  |            ^^^^^^^^^ ------------------
help: add missing generic argument
  |
3 | pub struct Magnitude<BYTES><const BYTES: usize>([u8; BYTES]);
  |                     +++++++

For more information about this error, try `rustc --explain E0107`.
error: could not compile `magnitude` (lib) due to 3 previous errors
```

Considering how inane the suggestion from rustc is, I suspect this may actually be a compiler bug, but I want to get a second opinion before I go opening an issue claiming a compiler error, especially for code involving macros I didn't write.

--- a/derive/src/lib.rs
+++ b/derive/src/lib.rs
@@ -377,7 +377,7 @@ pub fn derive_contiguous(
 /// also does not implement `StructuralPartialEq` / `StructuralEq` like
 /// `PartialEq` / `Eq` would. This means you can't pattern match on the values.
 ///
-/// ## Example
+/// ## Examples
 ///
 /// ```rust
 /// # use bytemuck_derive::{ByteEq, NoUninit};
@@ -389,6 +389,17 @@ pub fn derive_contiguous(
 ///   c: f32,
 /// }
 /// ```
+///
+/// ```rust
+/// # use bytemuck_derive::ByteEq;
+/// # use bytemuck::NoUninit;
+/// #[derive(Copy, Clone, ByteEq)]
+/// #[repr(C)]
+/// struct Test<const N: usize> {
+///   a: [u32; N],
+/// }
+/// unsafe impl<const N: usize> NoUninit for Test<N> {}
+/// ```
 #[proc_macro_derive(ByteEq)]
 pub fn derive_byte_eq(
   input: proc_macro::TokenStream,
@@ -396,16 +407,17 @@ pub fn derive_byte_eq(
   let input = parse_macro_input!(input as DeriveInput);
   let crate_name = bytemuck_crate_name(&input);
   let ident = input.ident;
+  let (impl_generics, ty_generics, where_clause) = input.generics.split_for_impl();
 
   proc_macro::TokenStream::from(quote! {
-    impl ::core::cmp::PartialEq for #ident {
+    impl #impl_generics ::core::cmp::PartialEq for #ident #ty_generics #where_clause {
       #[inline]
       #[must_use]
       fn eq(&self, other: &Self) -> bool {
         #crate_name::bytes_of(self) == #crate_name::bytes_of(other)
       }
     }
-    impl ::core::cmp::Eq for #ident { }
+    impl #impl_generics ::core::cmp::Eq for #ident #ty_generics #where_clause { }
   })
 }
 
@@ -418,7 +430,7 @@ pub fn derive_byte_eq(
 ///
 /// The hash does not match the standard library's `Hash` derive.
 ///
-/// ## Example
+/// ## Examples
 ///
 /// ```rust
 /// # use bytemuck_derive::{ByteHash, NoUninit};
@@ -430,6 +442,17 @@ pub fn derive_byte_eq(
 ///   c: f32,
 /// }
 /// ```
+///
+/// ```rust
+/// # use bytemuck_derive::ByteHash;
+/// # use bytemuck::NoUninit;
+/// #[derive(Copy, Clone, ByteHash)]
+/// #[repr(C)]
+/// struct Test<const N: usize> {
+///   a: [u32; N],
+/// }
+/// unsafe impl<const N: usize> NoUninit for Test<N> {}
+/// ```
 #[proc_macro_derive(ByteHash)]
 pub fn derive_byte_hash(
   input: proc_macro::TokenStream,
@@ -437,9 +460,10 @@ pub fn derive_byte_hash(
   let input = parse_macro_input!(input as DeriveInput);
   let crate_name = bytemuck_crate_name(&input);
   let ident = input.ident;
+  let (impl_generics, ty_generics, where_clause) = input.generics.split_for_impl();
 
   proc_macro::TokenStream::from(quote! {
-    impl ::core::hash::Hash for #ident {
+    impl #impl_generics ::core::hash::Hash for #ident #ty_generics #where_clause {
       #[inline]
       fn hash<H: ::core::hash::Hasher>(&self, state: &mut H) {
         ::core::hash::Hash::hash_slice(#crate_name::bytes_of(self), state)

